File: LeetCode/Medium/0133-clone-graph/0133-clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
    if node == nil { return node }
    
    ret := make([]*Node, 101)
    visited := make([]bool, 101)
    stack := make([]*Node, 1)
    ret[1] = &Node{1, make([]*Node, 0)}
    visited[1] = true
    stack[0] = node
    
    for len(stack) != 0 {
        temp := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        
        for _, neighbor := range temp.Neighbors {
            if !visited[neighbor.Val] {
                visited[neighbor.Val] = true
                ret[neighbor.Val] = &Node{neighbor.Val, make([]*Node, 0)}
                stack = append(stack, neighbor)
            }
            ret[temp.Val].Neighbors = append(ret[temp.Val].Neighbors, ret[neighbor.Val])
        }
    }
    
    return ret[1]
}

File: LeetCode/Medium/0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return
        
        ret = {}

        ret[1] = Node(node.val)
        next_stack = [node]

        while next_stack:
            temp = next_stack.pop()
            for v in temp.neighbors:
                if v.val not in ret:
                    ret[v.val] = Node(v.val)
                    next_stack.append(v)
                ret[temp.val].neighbors.append(ret[v.val]) 
            
        return ret[1]